airdrop-machine/sources/airdrop_machine.move

```diff
@@ -0,0 +1,289 @@
+module airdrop_machine::airdrop_machine {
+    use std::error;
+    use std::option;
+    use std::option::Option;
+    use std::signer;
+    use std::string::{String, utf8};
+
+    use aptos_framework::object::{Self, Object};
+
+    use aptos_token_objects::collection;
+    use aptos_token_objects::collection::Collection;
+    use aptos_token_objects::royalty;
+    use aptos_token_objects::royalty::Royalty;
+    use aptos_token_objects::token;
+    use aptos_token_objects::token::Token;
+
+    use minter::collection_components;
+    use minter::collection_properties;
+    use minter::token_components;
+    use minter::transfer_token;
+
+    /// The provided signer is not the collection owner.
+    const ENOT_OWNER: u64 = 1;
+    /// CollectionConfig resource does not exist in the object address.
+    const ECOLLECTION_CONFIG_DOES_NOT_EXIST: u64 = 2;
+    /// Royalty configuration is invalid because either numerator and denominator should be none or not none.
+    const EROYALTY_CONFIGURATION_INVALID: u64 = 3;
+    /// Token Minting has not yet started.
+    const EMINTING_HAS_NOT_STARTED: u64 = 7;
+
+    struct CollectionConfig has key {
+        extend_ref: object::ExtendRef,
+        ready_to_mint: bool,
+    }
+
+    public entry fun create_collection(
+        admin: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        mutable_collection_metadata: bool, // including description, uri, royalty, to make admin life easier
+        mutable_token_metadata: bool, // including description, name, properties, uri, to make admin life easier
+        tokens_burnable_by_collection_owner: bool,
+        tokens_transferrable_by_collection_owner: bool,
+        max_supply: Option<u64>,
+        royalty_numerator: Option<u64>,
+        royalty_denominator: Option<u64>,
+    ) {
+        create_collection_impl(
+            admin,
+            description,
+            name,
+            uri,
+            mutable_collection_metadata,
+            mutable_token_metadata,
+            tokens_burnable_by_collection_owner,
+            tokens_transferrable_by_collection_owner,
+            max_supply,
+            royalty_numerator,
+            royalty_denominator,
+        );
+    }
+
+    public entry fun mint(
+        user: &signer,
+        name: String,
+        description: String,
+        uri: String,
```

Comment:
hmnm, I think these will be fixed for the whole collection, I don't think we want users to be able to set these.

(and let's rename name to name_prefix, so it's clear index will be appended to it)

in case we have multiple assets, we should probably have a vector of uri, and since we don't have onchain randomness pick based on transaction_context::get_transaction_hash() / generate_auid_address()

Reply:
im hoping this contract can be used in a more generic way? so that for other airdrops, its applicable as well. so prob we can do these custom stuffs off-chain instead?

Reply:
The users shouldn't be passing these details in though, as they can pass whatever. Instead, we should have a config of these fields, and have the contract only mint from that?